src/microbe_stage/MicrobeAI.cs

```diff
@@ -370,43 +384,86 @@ private void RunAndTumble(Random random)
         microbe.State = Microbe.MicrobeState.Normal;
 
         var usefulCompounds = microbe.TotalAbsorbedCompounds.Where(x => microbe.Compounds.IsUseful(x.Key));
+        var compoundsWeights = ComputeCompoundsSearchWeights(usefulCompounds.ToList());
 
-        // If this microbe lacks glucose, don't bother with ammonia and phosphorous
-        // This algorithm doesn't try to determine if iron and sulfuric acid is useful to this microbe
-        if (microbe.Compounds.GetCompoundAmount(glucose) < 0.5f)
+        float gradientValue = 0.0f;
+        foreach (var compoundWeight in compoundsWeights.ToList())
         {
-            usefulCompounds = usefulCompounds.Where(x => x.Key != ammonia && x.Key != phosphates);
-        }
+            // Note this is about absorbed quantities (which is all microbe has access to), not real ones
+            float compoundDifference = 0.0f;
 
-        float compoundDifference = 0.0f;
-        foreach (var compound in usefulCompounds)
-        {
-            compoundDifference += compound.Value;
+            // No need to add a difference if compound was not absorbed
+            if (microbe.TotalAbsorbedCompounds.ContainsKey(compoundWeight.Key))
```

Comment:
Use try get value instead of checking for key existance and then looking up the value separately.